src/eos-updater-poll.c

```diff
@@ -203,6 +203,192 @@ get_fetchers (SourcesConfig *config,
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_VOLUME:
+          /* TODO: How to make this one testable? */
+          g_ptr_array_add (finders, ostree_repo_finder_mount_new (NULL));
+          break;
+
+        default:
+          g_assert_not_reached ();
+        }
+    }
+
+  g_ptr_array_add (finders, NULL);  /* NULL terminator */
+
+  /* TODO: Stop this at some point; think of a better way to store it and
+   * control its lifecycle. */
+  if (finder_avahi != NULL)
+    ostree_repo_finder_avahi_start (OSTREE_REPO_FINDER_AVAHI (finder_avahi),
+                                    &local_error);
+
+  if (local_error != NULL)
+    {
+      g_warning ("Avahi finder failed; removing it: %s", local_error->message);
+      g_ptr_array_remove (finders, finder_avahi);
+      g_clear_object (&finder_avahi);
+      g_clear_error (&local_error);
+    }
+
+  if (out_finder_avahi != NULL)
+    *out_finder_avahi = g_steal_pointer (&finder_avahi);
+
+  return g_steal_pointer (&finders);
+}
+
+static void
+async_result_cb (GObject      *source_object,
+                 GAsyncResult *result,
+                 gpointer      user_data)
+{
+  GAsyncResult **result_out = user_data;
+
+  *result_out = g_object_ref (result);
+}
+
+static EosUpdateInfo *
+metadata_fetch_new (OstreeRepo    *repo,
+                    SourcesConfig *config,
+                    GMainContext  *context,
+                    GCancellable  *cancellable,
+                    GError       **error)
+{
+  g_auto(OstreeRepoFinderResultv) results = NULL;
+  g_autoptr(EosUpdateInfo) info = NULL;
+  g_autofree gchar *checksum = NULL;
+  g_autoptr(GVariant) commit = NULL;
+  g_autofree gchar *booted_refspec = NULL, *new_refspec = NULL;
+  g_autoptr(OstreeCollectionRef) booted_collection_ref = NULL, new_collection_ref = NULL;
+  const gchar *upgrade_refspec;
+  const OstreeCollectionRef *upgrade_collection_ref;
+  g_autoptr(GPtrArray) finders = NULL;  /* (element-type OstreeRepoFinder) */
+  g_autoptr(OstreeRepoFinderAvahi) finder_avahi = NULL;
+
+  if (!get_booted_refspec (&booted_refspec, NULL, NULL, &booted_collection_ref, error))
+    return NULL;
+
+  if (booted_collection_ref == NULL)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED,
+                   "No collection ID set for currently booted deployment.");
+      return NULL;
+    }
+
+  upgrade_collection_ref = booted_collection_ref;
+  upgrade_refspec = booted_refspec;
+
+  finders = get_finders (config, context, &finder_avahi);
+
+  /* Check whether the commit is a redirection; if so, fetch the new ref and
+   * check again. */
+  do
+    {
+      const OstreeCollectionRef *refs[] = { NULL, NULL };
+      g_autoptr(GVariant) pull_options = NULL;
+      g_autoptr(GAsyncResult) find_result = NULL, pull_result = NULL;
+      g_auto(GVariantBuilder) builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("a{sv}"));
+
+      refs[0] = upgrade_collection_ref;
```

Comment:
Can we initialise `refs` directly with `{ upgrade_collection_ref, NULL }`?

Reply:
Indeed, this was left over from an earlier iteration of the code where that wasn’t possible. Fixup pushed.